crates/sdk/src/masp.rs

```diff
@@ -2038,181 +1942,493 @@ impl<U: ShieldedUtils + MaybeSend + MaybeSync> ShieldedContext<U> {
-                    for response_tx in txs {
-                        let height = BlockHeight(response_tx.height.value());
-                        let idx = TxIndex(response_tx.index);
-                        // Only process yet unprocessed transactions which have
-                        // been accepted by node VPs
-                        let should_process = !transfers
-                            .contains_key(&IndexedTx { height, index: idx })
-                            && block_results[u64::from(height) as usize]
-                                .is_accepted(idx.0 as usize);
-                        if !should_process {
-                            continue;
-                        }
-                        let tx = Tx::try_from(response_tx.tx.as_ref())
-                            .map_err(|e| Error::Other(e.to_string()))?;
-                        let mut wrapper = None;
-                        let mut transfer = None;
-                        extract_payload(tx, &mut wrapper, &mut transfer)?;
-                        // Epoch data is not needed for transparent transactions
-                        let epoch =
-                            wrapper.map(|x| x.epoch).unwrap_or_default();
-                        if let Some(transfer) = transfer {
-                            // Skip MASP addresses as they are already handled
-                            // by ShieldedContext
-                            if transfer.source == MASP
-                                || transfer.target == MASP
-                            {
-                                continue;
-                            }
-                            // Describe how a Transfer simply subtracts from one
-                            // account and adds the same to another
+                tx_ctx.scan_tx(*indexed_tx, *epoch, tx, stx)?;
+                self.unscanned.txs.remove(indexed_tx);
+            }
+            // Merge the context data originating from the unknown keys into the
+            // current context
+            self.merge(tx_ctx);
+        } else {
+            let resume_point =
+                std::cmp::max(self.latest_unscanned(), self.last_fetched);
+            // fetch only new transactions
+            let fetched = self
+                .fetch_shielded_transfers(
+                    client,
+                    io,
+                    resume_point,
+                    last_query_height,
+                )
+                .await?;
+            self.unscanned.txs.extend(fetched);
+            display_line!(io, "");
+        }
+        // Now that we possess the unspent notes corresponding to both old and
+        // new keys up until tx_pos, proceed to scan the new transactions.
+        let mut txs = BTreeMap::new();
+        std::mem::swap(&mut self.unscanned.txs, &mut txs);
+        let txs = ProgressLogging::new(txs, io, ProgressType::Scan);
+        for (indexed_tx, (epoch, tx, stx)) in txs {
+            if self.interrupted() {
```

Comment:
If an interrupt happens here, would `self.unscanned.txs` be left in a state where it's missing the very first transactions? If this were the case, would this hamper the client's future ability to scan the very first transactions with viewing keys that are not yet known at this point?

Reply:
Yes, the cache is cleared out. If new keys appear later, we have to repopoulate this cache.